File: src/aerialpod/db/repo.py
"""Typed query functions. All reads use the calling thread's connection;
multi-statement writes happen inside db.transaction() at the call site or here.
"""

from __future__ import annotations

import json
import time
from typing import Any

from . import connection, transaction
from .models import Episode, Podcast, QueueItem, from_row

# ---------------------------------------------------------------- app_state

DEFAULTS: dict[str, Any] = {
    "global_speed": 1.0,
    "volume": 1.0,  # linear QAudioOutput volume, app-level (PipeWire stream)
    "skip_fwd_secs": 30,
    "skip_back_secs": 10,
    "download_ahead_n": 1,
    "auto_add_to_queue": True,
    "auto_queue_position": "back",  # 'front' puts new episodes under the playing one
    "theme_mode": "system",  # 'system'|'light'|'dark'
    "accent": "#3584e4",
    "home_sections": ["queue", "continue", "inbox", "subscriptions"],
    "speed_presets": [1.0, 1.25, 1.5, 1.75, 2.0],
    "sync_interval_mins": 30,
    "media_next_action": "seek",  # keyboard/MPRIS ⏭⏮: 'seek' (ad-skip) | 'episode'
    "audio_device_mode": "follow_default",  # or 'pinned'
    "audio_device_id": None,
    "audio_device_description": None,
    "lan_sync_enabled": True,
    "lan_port": 47741,
    "lan_scan_subnets": True,  # unicast sweep of our own subnets (see lan.discovery)
}


def lan_device_id() -> str:
    """Stable identity for this install on the LAN mesh.

    Deliberately not the gpodder device id: that one is user-visible on
    gpodder.net and can be renamed there, while this must stay stable for
    peer bookkeeping and last-writer-wins tie-breaks.
    """
    did = get_state("lan_device_id")
    if not did:
        import uuid

        did = uuid.uuid4().hex
        set_state("lan_device_id", did)
    return did


def get_state(key: str, default: Any = None) -> Any:
    row = connection().execute("SELECT value FROM app_state WHERE key=?", (key,)).fetchone()
    if row is None:
        return DEFAULTS.get(key, default) if default is None else default
    return json.loads(row["value"])


def set_state(key: str, value: Any) -> None:
    connection().execute(
        "INSERT INTO app_state(key, value) VALUES(?, ?) "
        "ON CONFLICT(key) DO UPDATE SET value=excluded.value",
        (key, json.dumps(value)),
    )
    connection().commit()


# ---------------------------------------------------------------- podcasts


def all_podcasts(subscribed_only: bool = True) -> list[Podcast]:
    sql = "SELECT * FROM podcasts"
    if subscribed_only:
        sql += " WHERE subscribed=1"
    sql += " ORDER BY title COLLATE NOCASE"
    return [from_row(Podcast, r) for r in connection().execute(sql)]


def podcast_by_id(pid: int) -> Podcast | None:
    row = connection().execute("SELECT * FROM podcasts WHERE id=?", (pid,)).fetchone()
    return from_row(Podcast, row) if row else None


def podcast_by_feed_url(url: str) -> Podcast | None:
    row = connection().execute("SELECT * FROM podcasts WHERE feed_url=?", (url,)).fetchone()
    return from_row(Podcast, row) if row else None


def upsert_podcast(feed_url: str, *, sync_state: str = "add_pending", subscribed: int = 1) -> int:
    """Insert (or resubscribe) a podcast; returns its id."""
    with transaction() as conn:
        row = conn.execute("SELECT id FROM podcasts WHERE feed_url=?", (feed_url,)).fetchone()
        if row:
            conn.execute(
                "UPDATE podcasts SET subscribed=?, sync_state=? WHERE id=?",
                (subscribed, sync_state, row["id"]),
            )
            return row["id"]
        cur = conn.execute(
            "INSERT INTO podcasts(feed_url, sync_state, subscribed, added_at) VALUES(?,?,?,?)",
            (feed_url, sync_state, subscribed, int(time.time())),
        )
        return cur.lastrowid


def update_podcast_meta(pid: int, **cols: Any) -> None:
    if not cols:
        return
    sets = ", ".join(f"{k}=?" for k in cols)
    connection().execute(f"UPDATE podcasts SET {sets} WHERE id=?", (*cols.values(), pid))
    connection().commit()


def unsubscribe_podcast(pid: int) -> None:
    with transaction() as conn:
        conn.execute(
            "UPDATE podcasts SET subscribed=0, sync_state='remove_pending' WHERE id=?", (pid,)
        )
        conn.execute(
            "DELETE FROM queue WHERE episode_id IN (SELECT id FROM episodes WHERE podcast_id=?)",
            (pid,),
        )


def rewrite_feed_url(old: str, new: str) -> None:
    """Apply a gpodder update_urls rewrite. Keeps action matching working."""
    connection().execute("UPDATE podcasts SET feed_url=? WHERE feed_url=?", (new, old))
    connection().commit()


# ---------------------------------------------------------------- podcast settings


def podcast_settings(pid: int) -> dict[str, Any]:
    row = connection().execute(
        "SELECT * FROM podcast_settings WHERE podcast_id=?", (pid,)
    ).fetchone()
    return dict(row) if row else {}


SETTING_KEYS = (
    "custom_title",
    "playback_speed",
    "skip_intro_secs",
    "skip_outro_secs",
    "auto_add_to_queue",
    "auto_queue_position",
)


def set_podcast_setting(pid: int, key: str, value: Any) -> None:
    assert key in SETTING_KEYS
    connection().execute(
        f"INSERT INTO podcast_settings(podcast_id, {key}, updated_at, updated_by) "
        f"VALUES(?, ?, ?, ?) "
        f"ON CONFLICT(podcast_id) DO UPDATE SET {key}=excluded.{key}, "
        f"updated_at=excluded.updated_at, updated_by=excluded.updated_by",
        (pid, value, int(time.time()), lan_device_id()),
    )
    connection().commit()


def effective_auto_add(pid: int) -> bool:
    s = podcast_settings(pid).get("auto_add_to_queue")
    if s is None:
        return bool(get_state("auto_add_to_queue"))
    return bool(s)


def effective_queue_position(pid: int) -> str:
    """'front' or 'back' — where this podcast's new episodes enter the queue."""
    value = podcast_settings(pid).get("auto_queue_position")
    if value not in ("front", "back"):
        value = get_state("auto_queue_position")
    return "front" if value == "front" else "back"


def effective_speed(pid: int) -> float:
    s = podcast_settings(pid).get("playback_speed")
    return float(s) if s else float(get_state("global_speed"))


def display_title(p: Podcast) -> str:
    custom = podcast_settings(p.id).get("custom_title")
    return custom or p.title or p.feed_url


# ---------------------------------------------------------------- episodes


def episode_by_id(eid: int) -> Episode | None:
    row = connection().execute("SELECT * FROM episodes WHERE id=?", (eid,)).fetchone()
    return from_row(Episode, row) if row else None


def episodes_for_podcast(pid: int, limit: int = 500) -> list[Episode]:
    rows = connection().execute(
        "SELECT * FROM episodes WHERE podcast_id=? ORDER BY pub_date DESC LIMIT ?",
        (pid, limit),
    )
    return [from_row(Episode, r) for r in rows]


def update_episode(eid: int, **cols: Any) -> None:
    if not cols:
        return
    sets = ", ".join(f"{k}=?" for k in cols)
    connection().execute(f"UPDATE episodes SET {sets} WHERE id=?", (*cols.values(), eid))
    connection().commit()


def add_alias(eid: int, url: str) -> None:
    connection().execute(
        "INSERT OR IGNORE INTO episode_url_aliases(episode_id, url) VALUES(?, ?)", (eid, url)
    )
    connection().commit()


def in_progress_episodes(limit: int = 50) -> list[Episode]:
    rows = connection().execute(
        "SELECT e.* FROM episodes e JOIN podcasts p ON p.id=e.podcast_id "
        "WHERE p.subscribed=1 AND e.state != 'played' AND e.position_secs > 0 "
        "ORDER BY e.position_updated_at DESC LIMIT ?",
        (limit,),
    )
    return [from_row(Episode, r) for r in rows]


def inbox_episodes(limit: int = 100) -> list[Episode]:
    rows = connection().execute(
        "SELECT e.* FROM episodes e JOIN podcasts p ON p.id=e.podcast_id "
        "WHERE p.subscribed=1 AND e.state IN ('new','inbox') "
        "ORDER BY e.pub_date DESC LIMIT ?",
        (limit,),
    )
    return [from_row(Episode, r) for r in rows]


# ---------------------------------------------------------------- queue

GAP = 1024


def queue_episode_ids() -> set[int]:
    """One query for 'is this episode queued?' checks across a whole list."""
    return {r[0] for r in connection().execute("SELECT episode_id FROM queue")}


def podcast_display_info() -> dict[int, tuple[str, str | None]]:
    """Bulk {podcast_id: (display_title, image_url)} — avoids per-row queries
    when rendering episode lists."""
    rows = connection().execute(
        "SELECT p.id, COALESCE(s.custom_title, p.title, p.feed_url) AS title, p.image_url "
        "FROM podcasts p LEFT JOIN podcast_settings s ON s.podcast_id = p.id"
    )
    return {r["id"]: (r["title"], r["image_url"]) for r in rows}


def queue_items() -> list[QueueItem]:
    rows = connection().execute("SELECT * FROM queue ORDER BY position")
    return [from_row(QueueItem, r) for r in rows]


def queue_episodes() -> list[Episode]:
    rows = connection().execute(
        "SELECT e.* FROM queue q JOIN episodes e ON e.id=q.episode_id ORDER BY q.position"
    )
    return [from_row(Episode, r) for r in rows]


# ---------------------------------------------------------------- queue intent
#
# The replicated half of the queue: what the *user* decided, as opposed to
# what reconcile() derived. Written by QueueManager's user-facing ops only.


def record_intent(
    conn,
    episode_id: int,
    intent: str,
    *,
    position: int = 0,
    pinned: int = 0,
    origin: str = "manual",
    updated_at: int | None = None,
    updated_by: str | None = None,
) -> None:
    """Upsert one intent row. Takes an explicit connection so callers can fold
    it into the transaction that changes the queue itself."""
    assert intent in ("queued", "excluded")
    conn.execute(
        "INSERT INTO queue_intent(episode_id, intent, position, pinned, origin, "
        "updated_at, updated_by) VALUES(?,?,?,?,?,?,?) "
        "ON CONFLICT(episode_id) DO UPDATE SET intent=excluded.intent, "
        "position=excluded.position, pinned=excluded.pinned, origin=excluded.origin, "
        "updated_at=excluded.updated_at, updated_by=excluded.updated_by",
        (
            episode_id,
            intent,
            position,
            pinned,
            origin,
            updated_at if updated_at is not None else int(time.time()),
            updated_by or lan_device_id(),
        ),
    )


def queue_intents() -> list[dict[str, Any]]:
    rows = connection().execute("SELECT * FROM queue_intent ORDER BY position")
    return [dict(r) for r in rows]


def intent_for(eid: int) -> dict[str, Any] | None:
    row = connection().execute(
        "SELECT * FROM queue_intent WHERE episode_id=?", (eid,)
    ).fetchone()
    return dict(row) if row else None


def drop_intent(conn, episode_id: int) -> None:
    conn.execute("DELETE FROM queue_intent WHERE episode_id=?", (episode_id,))


def prune_intents(max_age_secs: int = 90 * 86400) -> int:
    """Drop intents that can no longer change anything: the episode is played
    and the decision is old enough that no peer is still catching up."""
    cutoff = int(time.time()) - max_age_secs
    with transaction() as conn:
        cur = conn.execute(
            "DELETE FROM queue_intent WHERE updated_at < ? AND episode_id IN "
            "(SELECT id FROM episodes WHERE state='played')",
            (cutoff,),
        )
        return cur.rowcount


def is_excluded(eid: int) -> bool:
    return (
        connection()
        .execute("SELECT 1 FROM queue_exclusions WHERE episode_id=?", (eid,))
        .fetchone()
        is not None
    )


# ---------------------------------------------------------------- outbox


def enqueue_action(
    podcast_url: str,
    episode_url: str,
    action: str,
    timestamp: str,
    started: int | None = None,
    position: int | None = None,
    total: int | None = None,
) -> None:
    connection().execute(
        "INSERT INTO action_outbox(podcast_url, episode_url, action, timestamp, started, position, total) "
        "VALUES(?,?,?,?,?,?,?)",
        (podcast_url, episode_url, action, timestamp, started, position, total),
    )
    connection().commit()


def outbox_actions() -> list[dict[str, Any]]:
    rows = connection().execute("SELECT * FROM action_outbox ORDER BY id")
    return [dict(r) for r in rows]


def clear_outbox(upto_id: int) -> None:
    connection().execute("DELETE FROM action_outbox WHERE id<=?", (upto_id,))
    connection().commit()


# ---------------------------------------------------------------- unmatched


def log_unmatched(podcast_url: str, episode_url: str, action: str, timestamp: str, payload: str) -> None:
    connection().execute(
        "INSERT INTO unmatched_actions(podcast_url, episode_url, action, timestamp, payload, received_at) "
        "VALUES(?,?,?,?,?,?)",
        (podcast_url, episode_url, action, timestamp, payload, int(time.time())),
    )
    connection().commit()


def unmatched_count() -> int:
    return connection().execute("SELECT COUNT(*) FROM unmatched_actions").fetchone()[0]


# ---------------------------------------------------------------- LAN peers


def known_peers() -> list[dict[str, Any]]:
    """Peers we have authenticated with before, most recently seen first."""
    rows = connection().execute("SELECT * FROM lan_peers ORDER BY last_seen DESC")
    return [dict(r) for r in rows]


def remember_peer(device_id: str, caption: str, address: str, port: int) -> None:
    connection().execute(
        "INSERT INTO lan_peers(device_id, caption, address, port, last_seen) "
        "VALUES(?,?,?,?,?) ON CONFLICT(device_id) DO UPDATE SET "
        "caption=excluded.caption, address=excluded.address, port=excluded.port, "
        "last_seen=excluded.last_seen",
        (device_id, caption, address, port, int(time.time())),
    )
    connection().commit()


def forget_peer(device_id: str) -> None:
    connection().execute("DELETE FROM lan_peers WHERE device_id=?", (device_id,))
    connection().commit()


def manual_peers() -> list[tuple[str, int]]:
    rows = connection().execute(
        "SELECT address, port FROM lan_manual_peers ORDER BY address"
    )
    return [(r["address"], r["port"]) for r in rows]


def add_manual_peer(address: str, port: int) -> None:
    connection().execute(
        "INSERT OR IGNORE INTO lan_manual_peers(address, port) VALUES(?,?)",
        (address, port),
    )
    connection().commit()


def remove_manual_peer(address: str, port: int) -> None:
    connection().execute(
        "DELETE FROM lan_manual_peers WHERE address=? AND port=?", (address, port)
    )
    connection().commit()

File: mobile/shared/src/commonMain/kotlin/org/aerialpod/core/db/Repo.kt
package org.aerialpod.core.db

import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.booleanOrNull
import kotlinx.serialization.json.doubleOrNull
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.longOrNull
import org.aerialpod.core.epochSeconds
import org.aerialpod.core.iso8601Utc
import org.aerialpod.core.lan.secureRandomBytes
import org.aerialpod.core.lan.toHex

/**
 * Typed accessors over the database — the port of `db/repo.py` that the sync
 * layer needs.
 *
 * `app_state` holds JSON-encoded values exactly as the desktop writes them, so
 * a database copied between the two would still read correctly. That costs one
 * parse per lookup and buys a schema that does not fork.
 */
class Repo(
    val db: AerialPodDatabase,
    private val now: () -> Long = ::epochSeconds,
) {
    private val json = Json

    // ---------------------------------------------------------------- app_state

    fun rawState(key: String): String? =
        db.syncQueries.getState(key).executeAsOneOrNull()?.value_

    fun setState(key: String, value: String) {
        db.syncQueries.setState(key, json.encodeToString(JsonPrimitive.serializer(), JsonPrimitive(value)))
    }

    fun setState(key: String, value: Long) {
        db.syncQueries.setState(key, value.toString())
    }

    fun setState(key: String, value: Boolean) {
        db.syncQueries.setState(key, if (value) "true" else "false")
    }

    fun stateString(key: String, default: String): String =
        rawState(key)?.let { runCatching { json.parseToJsonElement(it) }.getOrNull() }
            ?.let { (it as? JsonPrimitive)?.contentOrNull } ?: default

    fun stateLong(key: String, default: Long): Long =
        rawState(key)?.let { runCatching { json.parseToJsonElement(it) }.getOrNull() }
            ?.let { (it as? JsonPrimitive)?.longOrNull } ?: default

    fun stateDouble(key: String, default: Double): Double =
        rawState(key)?.let { runCatching { json.parseToJsonElement(it) }.getOrNull() }
            ?.let { (it as? JsonPrimitive)?.doubleOrNull } ?: default

    /**
     * A JSON array of strings, which is how the desktop stores `home_sections`
     * and friends. Shape mismatches fall back rather than throw — the row is
     * shared with another program and editable by hand.
     */
    fun stateStringList(key: String, default: List<String>): List<String> {
        val raw = rawState(key) ?: return default
        val parsed = runCatching { json.parseToJsonElement(raw) }.getOrNull() as? JsonArray ?: return default
        return parsed.mapNotNull { (it as? JsonPrimitive)?.contentOrNull }
    }

    fun setState(key: String, value: List<String>) {
        db.syncQueries.setState(
            key,
            json.encodeToString(JsonArray.serializer(), JsonArray(value.map(::JsonPrimitive))),
        )
    }

    fun stateBool(key: String, default: Boolean): Boolean =
        rawState(key)?.let { runCatching { json.parseToJsonElement(it) }.getOrNull() }
            ?.let { (it as? JsonPrimitive)?.booleanOrNull } ?: default

    /**
     * Stable identity for this install on the peer mesh.
     *
     * Deliberately not the gpodder device id: that one is user-visible on
     * gpodder.net and can be renamed there, while this must stay stable for
     * peer bookkeeping and for the last-writer-wins tie-break — where two
     * devices comparing the same pair of ids have to reach the same answer.
     *
     * 16 random bytes as hex, which is the shape the desktop's `uuid4().hex`
     * produces, so neither side has to special-case the other's.
     */
    fun lanDeviceId(): String {
        rawState(LAN_DEVICE_ID)?.let { raw ->
            val existing = (runCatching { json.parseToJsonElement(raw) }.getOrNull()
                as? JsonPrimitive)?.contentOrNull
            if (!existing.isNullOrEmpty()) return existing
        }
        val generated = secureRandomBytes(16).toHex()
        setState(LAN_DEVICE_ID, generated)
        return generated
    }

    fun lanPort(): Int = stateLong(LAN_PORT, DEFAULT_LAN_PORT).toInt()

    fun lanSyncEnabled(): Boolean = stateBool(LAN_SYNC_ENABLED, true)

    // ---------------------------------------------------------------- podcasts

    fun podcastById(id: Long): Podcasts? =
        db.podcastsQueries.podcastById(id).executeAsOneOrNull()

    fun podcastByFeedUrl(url: String): Podcasts? =
        db.podcastsQueries.podcastByFeedUrl(url).executeAsOneOrNull()

    fun allPodcasts(): List<Podcasts> = db.podcastsQueries.selectAllPodcasts().executeAsList()

    fun subscribedPodcasts(): List<Podcasts> =
        db.podcastsQueries.selectSubscribed().executeAsList()

    /** Insert, or resubscribe an existing feed; returns its id. */
    fun upsertPodcast(
        feedUrl: String,
        syncState: String = "add_pending",
        subscribed: Long = 1,
    ): Long = db.transactionWithResult {
        val existing = podcastByFeedUrl(feedUrl)
        if (existing != null) {
            db.podcastsQueries.resubscribePodcast(subscribed, syncState, existing.id)
            existing.id
        } else {
            db.podcastsQueries.insertPodcast(feedUrl, syncState, subscribed, now())
            db.podcastsQueries.lastInsertId().executeAsOne()
        }
    }

    /**
     * Unsubscribe, and drop the podcast's episodes from the queue.
     *
     * The queue rows go immediately rather than waiting for reconcile: an
     * unsubscribed podcast has no claim on the queue, and leaving them until
     * the next pass would show the user episodes from a feed they just removed.
     */
    fun unsubscribePodcast(podcastId: Long) {
        db.transaction {
            db.podcastsQueries.unsubscribePodcast(podcastId)
            db.queueQueries.deleteQueueForPodcast(podcastId)
        }
    }

    // ---------------------------------------------------------------- episodes

    fun episodeById(id: Long): Episodes? =
        db.episodesQueries.episodeById(id).executeAsOneOrNull()

    /** Unplayed and unstarted, newest first — the Inbox. */
    fun inboxEpisodes(limit: Long = 200): List<Episodes> =
        db.episodesQueries.inboxEpisodes(limit).executeAsList()

    /** Started but not finished, most recently touched first. */
    fun inProgressEpisodes(limit: Long = 200): List<Episodes> =
        db.episodesQueries.inProgressEpisodes(limit).executeAsList()

    fun episodesForPodcast(podcastId: Long, limit: Long = 500): List<Episodes> =
        db.episodesQueries.episodesForPodcast(podcastId, limit).executeAsList()

    fun queueEpisodeIds(): Set<Long> =
        db.queueQueries.queueEpisodeIds().executeAsList().toSet()

    /**
     * `{podcast_id: (display title, cover)}` in one query.
     *
     * The desktop learned this the hard way — a per-row lookup while rendering
     * an episode list is what made opening a podcast hang.
     */
    fun podcastDisplayInfo(): Map<Long, PodcastDisplay> =
        db.podcastsQueries.podcastDisplayInfo().executeAsList()
            .associate { it.id to PodcastDisplay(it.id, it.title, it.image_url) }

    fun downloadedEpisodes(): List<Episodes> =
        db.episodesQueries.downloadedEpisodes().executeAsList()

    fun setDownloadState(episodeId: Long, state: String, path: String? = null) {
        db.episodesQueries.setDownloadState(path, state, episodeId)
    }

    /** Pin a download so the policy never evicts it. */
    fun setKeepDownload(episodeId: Long, keep: Boolean) {
        db.episodesQueries.setKeepDownload(if (keep) 1L else 0L, episodeId)
    }

    /**
     * How much of the queue to keep downloaded: a count, a fraction like
     * `1/4`, or `all`.
     *
     * A deliberate mobile-only extension. The desktop's `download_ahead_n` is
     * an integer and stays one — a fraction written into that key would crash
     * its `int()` — so this lives beside it and falls back to it, which keeps
     * an upgraded install on whatever it was already set to.
     */
    fun downloadAhead(): String =
        rawState(DOWNLOAD_AHEAD)?.let { stateString(DOWNLOAD_AHEAD, "") }?.takeIf { it.isNotEmpty() }
            ?: stateLong(DOWNLOAD_AHEAD_N, 1).toString()

    fun setDownloadAhead(value: String) {
        setState(DOWNLOAD_AHEAD, value)
        // Keep the desktop's key readable and roughly right, so a database that
        // ever meets the desktop does not hand it something it cannot parse.
        setState(DOWNLOAD_AHEAD_N, value.toLongOrNull() ?: 1L)
    }

    fun addAlias(episodeId: Long, url: String) {
        db.episodesQueries.addAlias(episodeId, url)
    }

    // ---------------------------------------------------------------- intent

    /**
     * Upsert one intent row — the replicated half of the queue.
     *
     * `updatedAt`/`updatedBy` are parameters rather than always-now because a
     * merge has to preserve the *peer's* stamp: re-stamping an incoming record
     * with our own clock would make it look locally authored and win every
     * subsequent conflict against the device that actually wrote it.
     */
    fun recordIntent(
        episodeId: Long,
        intent: String,
        position: Long = 0,
        pinned: Long = 0,
        origin: String = "manual",
        updatedAt: Long? = null,
        updatedBy: String? = null,
    ) {
        require(intent == "queued" || intent == "excluded") { "unknown intent $intent" }
        db.queueQueries.upsertIntent(
            episode_id = episodeId,
            intent = intent,
            position = position,
            pinned = pinned,
            origin = origin,
            updated_at = updatedAt ?: now(),
            updated_by = updatedBy ?: lanDeviceId(),
        )
    }

    fun dropIntent(episodeId: Long) {
        db.queueQueries.dropIntent(episodeId)
    }

    /**
     * Drop intents that can no longer change anything: the episode is played
     * and the decision is old enough that no peer is still catching up.
     */
    fun pruneIntents(maxAgeSecs: Long = 90L * 86_400): Long {
        val cutoff = now() - maxAgeSecs
        val doomed = db.queueQueries.countPrunableIntents(cutoff).executeAsOne()
        db.queueQueries.pruneIntents(cutoff)
        return doomed
    }

    // ---------------------------------------------------------------- settings

    fun podcastSettings(podcastId: Long): Podcast_settings? =
        db.podcastsQueries.settingsForPodcast(podcastId).executeAsOneOrNull()

    /**
     * Change one per-podcast setting, leaving the rest alone.
     *
     * Read-modify-write because the only upsert the schema offers writes every
     * column — which is what a snapshot merge wants (it carries the peer's
     * whole record) and exactly what a local edit must not do, or setting the
     * speed would silently blank the custom title. The read and the write share
     * a transaction so two edits in flight cannot lose one another.
     *
     * Stamps `updated_at`/`updated_by` with this device, which is what makes
     * the change win the next merge against an older copy.
     */
    fun updatePodcastSettings(podcastId: Long, mutate: (PodcastSettings) -> PodcastSettings) {
        db.transaction {
            val current = PodcastSettings.from(podcastSettings(podcastId))
            val next = mutate(current)
            db.podcastsQueries.upsertAllSettings(
                podcast_id = podcastId,
                custom_title = next.customTitle,
                playback_speed = next.playbackSpeed,
                skip_intro_secs = next.skipIntroSecs,
                skip_outro_secs = next.skipOutroSecs,
                auto_add_to_queue = next.autoAddToQueue,
                auto_queue_position = next.autoQueuePosition,
                updated_at = now(),
                updated_by = lanDeviceId(),
            )
        }
    }

    fun effectiveAutoAdd(podcastId: Long): Boolean =
        podcastSettings(podcastId)?.auto_add_to_queue?.let { it != 0L }
            ?: stateBool(AUTO_ADD_TO_QUEUE, true)

    /** 'front' or 'back' — where this podcast's new episodes enter the queue. */
    fun effectiveQueuePosition(podcastId: Long): String {
        val value = podcastSettings(podcastId)?.auto_queue_position
            ?.takeIf { it == "front" || it == "back" }
            ?: stateString(AUTO_QUEUE_POSITION, "back")
        return if (value == "front") "front" else "back"
    }

    fun effectiveSpeed(podcastId: Long): Double =
        podcastSettings(podcastId)?.playback_speed?.takeIf { it > 0.0 }
            ?: stateDouble(GLOBAL_SPEED, 1.0)

    fun displayTitle(podcast: Podcasts): String =
        podcastSettings(podcast.id)?.custom_title?.takeIf { it.isNotBlank() }
            ?: podcast.title?.takeIf { it.isNotBlank() }
            ?: podcast.feed_url

    // ---------------------------------------------------------------- queue reads

    fun queueItems(): List<Queue> = db.queueQueries.queueItems().executeAsList()

    fun queueEpisodes(): List<Episodes> = db.queueQueries.queueEpisodes().executeAsList()

    fun queueItem(episodeId: Long): Queue? =
        db.queueQueries.queueItemFor(episodeId).executeAsOneOrNull()

    fun isExcluded(episodeId: Long): Boolean =
        db.queueQueries.isExcluded(episodeId).executeAsOneOrNull() != null

    // ---------------------------------------------------------------- outbox

    /**
     * Queue a gpodder episode action for upload.
     *
     * This is the only channel that reaches the phone's other podcast apps —
     * peer sync carries the queue, but played/new state travels here, and
     * having both write it would just be a race.
     */
    fun enqueueAction(
        podcastUrl: String,
        episodeUrl: String,
        action: String,
        timestamp: String = iso8601Utc(now()),
        started: Long? = null,
        position: Long? = null,
        total: Long? = null,
    ) {
        db.syncQueries.enqueueAction(
            podcast_url = podcastUrl,
            episode_url = episodeUrl,
            action = action,
            timestamp = timestamp,
            started = started,
            position = position,
            total = total,
        )
    }

    fun outboxActions(): List<Action_outbox> = db.syncQueries.outboxActions().executeAsList()

    /**
     * Record an action we could not attach to a local episode.
     *
     * Kept rather than dropped because an unmatched action is the visible
     * symptom of the URL ladder failing on a real feed — it is the only way to
     * find out which CDN needs a new tracker pattern.
     */
    fun logUnmatched(
        podcastUrl: String,
        episodeUrl: String,
        action: String,
        timestamp: String,
        payload: String,
    ) {
        db.syncQueries.logUnmatched(podcastUrl, episodeUrl, action, timestamp, payload, now())
    }

    fun unmatchedCount(): Long = db.syncQueries.unmatchedCount().executeAsOne()

    fun clearOutbox(upToId: Long) {
        db.syncQueries.clearOutbox(upToId)
    }

    // ---------------------------------------------------------------- peers

    fun knownPeers(): List<Lan_peers> = db.lanPeersQueries.knownPeers().executeAsList()

    fun rememberPeer(deviceId: String, caption: String, address: String, port: Int) {
        db.lanPeersQueries.rememberPeer(deviceId, caption, address, port.toLong(), now())
    }

    fun forgetPeer(deviceId: String) {
        db.lanPeersQueries.forgetPeer(deviceId)
    }

    fun manualPeers(): List<Pair<String, Int>> =
        db.lanPeersQueries.manualPeers().executeAsList().map { it.address to it.port.toInt() }

    fun addManualPeer(address: String, port: Int) {
        db.lanPeersQueries.addManualPeer(address, port.toLong())
    }

    fun removeManualPeer(address: String, port: Int) {
        db.lanPeersQueries.removeManualPeer(address, port.toLong())
    }

    data class PodcastDisplay(val id: Long, val title: String, val imageUrl: String?)

    companion object {
        // The subset of the desktop's repo.DEFAULTS that the shared core reads.
        // UI-only defaults (theme, accent, home sections) belong to the app.
        const val GLOBAL_SPEED = "global_speed"
        const val SKIP_FWD_SECS = "skip_fwd_secs"
        const val SKIP_BACK_SECS = "skip_back_secs"
        const val DOWNLOAD_AHEAD_N = "download_ahead_n"
        const val DOWNLOAD_AHEAD = "download_ahead"
        const val AUTO_ADD_TO_QUEUE = "auto_add_to_queue"
        const val AUTO_QUEUE_POSITION = "auto_queue_position"
        const val SYNC_INTERVAL_MINS = "sync_interval_mins"
        const val LAN_DEVICE_ID = "lan_device_id"
        const val LAN_PORT = "lan_port"
        const val LAN_SYNC_ENABLED = "lan_sync_enabled"
        const val DEFAULT_LAN_PORT = 47722L
    }
}
